External/FEXCore/Source/Interface/Core/OpcodeDispatcher.h

```diff
@@ -557,23 +565,409 @@ enum class SelectionFlag {
-  void GenerateFlags_ADD(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2, bool UpdateCF = true);
-  void GenerateFlags_MUL(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *High);
-  void GenerateFlags_UMUL(FEXCore::X86Tables::DecodedOp Op, OrderedNode *High);
-  void GenerateFlags_Logical(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2);
-  void GenerateFlags_ShiftLeft(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2);
-  void GenerateFlags_ShiftLeftImmediate(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, uint64_t Shift);
-  void GenerateFlags_ShiftRight(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2);
-  void GenerateFlags_ShiftRightImmediate(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, uint64_t Shift);
-  void GenerateFlags_SignShiftRight(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2);
-  void GenerateFlags_SignShiftRightImmediate(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, uint64_t Shift);
-  void GenerateFlags_RotateRight(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2);
-  void GenerateFlags_RotateLeft(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, OrderedNode *Src2);
-  void GenerateFlags_RotateRightImmediate(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, uint64_t Shift);
-  void GenerateFlags_RotateLeftImmediate(FEXCore::X86Tables::DecodedOp Op, OrderedNode *Res, OrderedNode *Src1, uint64_t Shift);
+  enum class FlagsGenerationType {
+    TYPE_NONE,
+    TYPE_ADC,
+    TYPE_SBB,
+    TYPE_SUB,
+    TYPE_ADD,
+    TYPE_MUL,
+    TYPE_UMUL,
+    TYPE_LOGICAL,
+    TYPE_LSHL,
+    TYPE_LSHLI,
+    TYPE_LSHR,
+    TYPE_LSHRI,
+    TYPE_ASHR,
+    TYPE_ASHRI,
+    TYPE_ROR,
+    TYPE_RORI,
+    TYPE_ROL,
+    TYPE_ROLI,
+    TYPE_FCMP,
+  };
+
+  // Deferred flag generation tracking structure
+  // This structure is used to track RFlags from ALU ops for invalidation
+  // This tracks only six RFLAGS flags.
+  // Specifically: CF, PF, AF, ZF, SF, OF
+  //
+  // These six flags are heavily generated through basic ALU ops and balloon the IR if not early eliminated
+  // This tracking structure only tracks single blocks and requires RFLAGS canonicalization at block ender ops
+  // Some flags generating ALU ops only touch part of the registers, In these cases it will do canonicalization up front
+  // This means we still need our IR passes to eliminate all redundant flags accesses but this light OpcodeDispatcher optimization
+  // doesn't take it to that level
+  //
+  // Future ideas: Use an invalidation mask to do partial canonicalization for the instructions that don't do full invalidation
```

Comment:
With "full invalidation", are you referring to instructions that only set some but not all of these six RFLAGS flags? What are we doing about these instructions currently? Do we just not apply the optimization for those, or do we backtrack to the previous instruction for non-overridden flags?

Reply:
Instructions that do partial flag generation will calculate rflags up front then write to rflags directly.
Updated the comment to better reflect this.